File: src/main/java/com/morak/performancetracker/junit/PerformanceTrackerSetupExtension.java
package com.morak.performancetracker.junit;

import com.morak.performancetracker.PerformanceTracker;
import com.morak.performancetracker.context.Accumulator;
import com.morak.performancetracker.context.ContextManager;
import com.morak.performancetracker.context.TestMetadata;
import org.junit.jupiter.api.extension.*;
import org.springframework.context.ApplicationContext;
import org.springframework.test.context.junit.jupiter.SpringExtension;

public class PerformanceTrackerSetupExtension implements BeforeAllCallback, BeforeEachCallback, AfterEachCallback, AfterAllCallback {

    @Override
    public void beforeAll(ExtensionContext context) throws Exception {
        ApplicationContext applicationContext = getApplicationContext(context);
        ContextManager manager = applicationContext.getBean(determineContextManager(context));
        TestMetadata testMetadata = TestMetadata.of(context.getRequiredTestClass().getName());
        manager.beforeClass(testMetadata);
    }

    @Override
    public void beforeEach(ExtensionContext context) throws Exception {
        ApplicationContext applicationContext = getApplicationContext(context);
        ContextManager manager = applicationContext.getBean(determineContextManager(context));
        TestMetadata testMetadata = TestMetadata.of(context.getRequiredTestClass().getName(), context.getRequiredTestMethod().getName());
        manager.beforeEach(testMetadata);
    }

    @Override
    public void afterEach(ExtensionContext context) {
        ApplicationContext applicationContext = getApplicationContext(context);
        if (context.getExecutionException().isPresent()) {
            return;
        }
        ContextManager manager = applicationContext.getBean(determineContextManager(context));
        TestMetadata testMetadata = TestMetadata.of(context.getRequiredTestClass().getName(), context.getRequiredTestMethod().getName());
        manager.afterEach(applicationContext.getBean(Accumulator.class), testMetadata);
    }

    @Override
    public void afterAll(ExtensionContext context) {
        ApplicationContext applicationContext = getApplicationContext(context);
        ContextManager manager = applicationContext.getBean(determineContextManager(context));
        TestMetadata testMetadata = TestMetadata.of(context.getRequiredTestClass().getName());
        manager.afterClass(applicationContext.getBean(Accumulator.class), testMetadata);
    }

    private Class<? extends ContextManager> determineContextManager(ExtensionContext context) {
        PerformanceTracker annotation = context.getRequiredTestClass().getAnnotation(PerformanceTracker.class);
        return annotation.context().getContextManagerClass();
    }

    private ApplicationContext getApplicationContext(ExtensionContext context) {
        return SpringExtension.getApplicationContext(context);
    }
}

File: src/main/kotlin/com/morak/performancetracker/junit/PerformanceTrackerSetupExtension.kt
package com.morak.performancetracker.junit

import com.morak.performancetracker.PerformanceTracker
import com.morak.performancetracker.context.Accumulator
import com.morak.performancetracker.context.ContextManager
import com.morak.performancetracker.context.TestMetadata
import org.junit.jupiter.api.extension.*
import org.springframework.context.ApplicationContext
import org.springframework.test.context.junit.jupiter.SpringExtension

class PerformanceTrackerSetupExtension : BeforeAllCallback, BeforeEachCallback, AfterEachCallback, AfterAllCallback {

    override fun beforeAll(context: ExtensionContext) {
        val applicationContext = getApplicationContext(context)
        val manager = applicationContext.getBean(determineContextManager(context))
        val testMetadata = TestMetadata.of(context.requiredTestClass.name)
        manager.beforeClass(testMetadata)
    }

    override fun beforeEach(context: ExtensionContext) {
        val applicationContext = getApplicationContext(context)
        val manager = applicationContext.getBean(determineContextManager(context))
        val testMetadata = TestMetadata.of(context.requiredTestClass.name, context.requiredTestMethod.name)
        manager.beforeEach(testMetadata)
    }

    override fun afterEach(context: ExtensionContext) {
        if (context.executionException.isPresent) {
            return
        }

        val applicationContext = getApplicationContext(context)
        val manager = applicationContext.getBean(determineContextManager(context))
        val testMetadata = TestMetadata.of(context.requiredTestClass.name, context.requiredTestMethod.name)
        manager.afterEach(applicationContext.getBean(Accumulator::class.java), testMetadata)
    }

    override fun afterAll(context: ExtensionContext) {
        val applicationContext = getApplicationContext(context)
        val manager = applicationContext.getBean(determineContextManager(context))
        val testMetadata = TestMetadata.of(context.requiredTestClass.name)
        manager.afterClass(applicationContext.getBean(Accumulator::class.java), testMetadata)
    }

    private fun determineContextManager(context: ExtensionContext): Class<out ContextManager> {
        val annotation = context.requiredTestClass.getAnnotation(PerformanceTracker::class.java)
        return annotation.context.contextManagerClass.java
    }

    private fun getApplicationContext(context: ExtensionContext): ApplicationContext {
        return SpringExtension.getApplicationContext(context)
    }
}
